gf/include/alps/gf_new/tensors/tensor_base.h

```diff
@@ -0,0 +1,489 @@
+         *
+         * @param data  - pointer to the raw data buffer
+         * @param sizes - array with sizes for each dimension
+         */
+        tensor_base(T *data, std::array < size_t, D > sizes) : data_(viewType(data, size(sizes))), sizes_(sizes) {
+          fill_acc_sizes();
+        }
+
+        /**
+         * Create empty Tensor with provided sizes for each dimensions
+         *
+         * @tparam X type of storage to be created. Should always be DataStorage
+         * @param sizes - array of data dimensions
+         */
+        template<typename X = C>
+        tensor_base(typename std::enable_if < std::is_same < X, data_storage < T > >::value, const std::array < size_t, D > & >::type sizes) : data_(size(sizes)), sizes_(sizes) {
+          fill_acc_sizes();
+        }
+
+        explicit tensor_base(size_t size) : data_(size), sizes_{{size}} {
+          static_assert(1 == D, "Wrong dimension");
+          fill_acc_sizes();
+        }
+
+        /// copy constructor
+        tensor_base(const tensorType& rhs) : data_(rhs.data()), sizes_(rhs.sizes()), acc_sizes_(rhs.acc_sizes()) {}
+
+        template<typename St = C>
+        tensor_base(typename std::enable_if<std::is_same<St, viewType>::value, tensorType>::type& rhs) :
+          data_(rhs.data()), sizes_(rhs.sizes()), acc_sizes_(rhs.acc_sizes()) {}
+        /// move constructor
+        tensor_base(tensorType &&rhs) : data_(rhs.data()), sizes_(rhs.sizes()), acc_sizes_(rhs.acc_sizes()) {}
+        /// copy constructor
+        tensor_base(const tensorViewType &rhs) : data_(rhs.data()), sizes_(rhs.sizes()), acc_sizes_(rhs.acc_sizes()) {}
+        /// move constructor
+        tensor_base(tensorViewType &&rhs) : data_(rhs.data()), sizes_(rhs.sizes()), acc_sizes_(rhs.acc_sizes()) {}
+
+
+        /// Copy assignment
+        tensor_base < T, D, C > &operator=(const tensor_base < T, D, C > &rhs) {
+          data_ = rhs.data_;
+          sizes_ = rhs.sizes_;
+          acc_sizes_ = rhs.acc_sizes_;
+          return *this;
+        }
+
+        /// Move assignment
+        tensor_base < T, D, C > &operator=(tensor_base < T, D, C > &&rhs) noexcept {
+          data_ = rhs.data_;
+          sizes_ = rhs.sizes_;
+          acc_sizes_ = rhs.acc_sizes_;
+          return *this;
+        }
+
+        bool operator==(const tType& rhs) const {
+          return std::equal(sizes_.begin(), sizes_.end(), rhs.sizes_.begin()) && data_ == rhs.data_;
+        }
+
+        /**
+         * Get data point for the specific set of indices
+         * @tparam IndexTypes - types of tail indices
+         * @param t1          - head index
+         * @param indices     - tail indices
+         * @return value of tensor at the (t1, indices...) point
+         */
+        template<typename ...IndexTypes>
+        T operator()(typename std::enable_if < sizeof...(IndexTypes) == D - 1, size_t >::type t1, IndexTypes ... indices) const {
```

Comment:
I think a `static_assert` on `sizeof...(IndexTypes)` might be clearer/give better error messages here. I don't think you need the SFINAE to disambiguate the overload resolution here.

Reply:
nevermind you do need it to disambiguate with the slicing.